// @aws-amplify/cli/lib/lib/init-steps/preInitSetup.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var execSync = require('child_process').execSync;
var getPackageManager = require('../packageManagerHelpers').getPackageManager;
var normalizePackageManagerForOS = require('../packageManagerHelpers').normalizePackageManagerForOS;
var generateLocalEnvInfoFile = require('./s9-onSuccess').generateLocalEnvInfoFile;
var url = require('url');
var fs = require('fs');
function run(context) {
    return __awaiter(this, void 0, void 0, function () {
        var repoUrl;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!context.parameters.options.app) return [3 /*break*/, 5];
                    // Setting up a sample app
                    context.print.warning('Note: Amplify does not have knowledge of the url provided');
                    repoUrl = context.parameters.options.app;
                    return [4 /*yield*/, validateGithubRepo(context, repoUrl)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, cloneRepo(context, repoUrl)];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, installPackage()];
                case 3:
                    _a.sent();
                    return [4 /*yield*/, setLocalEnvDefaults(context)];
                case 4:
                    _a.sent();
                    _a.label = 5;
                case 5: return [2 /*return*/, context];
            }
        });
    });
}
/**
 * Checks whether a url is a valid remote github repository
 *
 * @param repoUrl the url to validated
 * @throws error if url is not a valid remote github url
 */
function validateGithubRepo(context, repoUrl) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            try {
                url.parse(repoUrl);
                execSync("git ls-remote " + repoUrl, { stdio: 'ignore' });
            }
            catch (e) {
                context.print.error('Invalid remote github url');
                process.exit(1);
            }
            return [2 /*return*/];
        });
    });
}
/**
 * Clones repo from url to current directory (must be empty)
 *
 * @param repoUrl the url to be cloned
 */
function cloneRepo(context, repoUrl) {
    return __awaiter(this, void 0, void 0, function () {
        var files;
        return __generator(this, function (_a) {
            files = fs.readdirSync(process.cwd());
            if (files.length > 0) {
                context.print.error('Please ensure you run this command in an empty directory');
                process.exit(1);
            }
            try {
                execSync("git clone " + repoUrl + " .", { stdio: 'inherit' });
            }
            catch (e) {
                process.exit(1);
            }
            return [2 /*return*/];
        });
    });
}
/**
 * Install package using the correct package manager if package handling file exists
 *
 * @param packageManager either npm or yarn
 */
function installPackage() {
    return __awaiter(this, void 0, void 0, function () {
        var packageManager, normalizedPackageManager;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getPackageManager()];
                case 1:
                    packageManager = _a.sent();
                    return [4 /*yield*/, normalizePackageManagerForOS(packageManager)];
                case 2:
                    normalizedPackageManager = _a.sent();
                    if (normalizedPackageManager) {
                        execSync(normalizedPackageManager + " install", { stdio: 'inherit' });
                    }
                    return [2 /*return*/];
            }
        });
    });
}
/**
 * Set the default environment and editor for the local env
 *
 * @param context
 */
function setLocalEnvDefaults(context) {
    return __awaiter(this, void 0, void 0, function () {
        var projectPath, defaultEditor, envName;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    projectPath = process.cwd();
                    defaultEditor = 'vscode';
                    envName = 'sampledev';
                    context.print.warning("Setting default editor to " + defaultEditor);
                    context.print.warning("Setting environment to " + envName);
                    context.print.warning('Run amplify configure project to change the default configuration later');
                    context.exeInfo.localEnvInfo = {
                        projectPath: projectPath,
                        defaultEditor: defaultEditor,
                        envName: envName,
                    };
                    context.exeInfo.inputParams.amplify.envName = envName;
                    return [4 /*yield*/, generateLocalEnvInfoFile(context)];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
module.exports = {
    run: run,
};
