// amplify-appsync-simulator/lib/velocity/util/time.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var moment = require("moment");
require("moment-timezone");
require("moment-jdateformatparser");
var parseTimestamp = function (dateTime, format, timezone) {
    if (!dateTime || !format) {
        return null;
    }
    try {
        var momentFormatString = moment().toMomentFormatString(format);
        return timezone ? moment.tz(dateTime, momentFormatString, timezone) : moment(dateTime, momentFormatString);
    }
    catch (e) {
        return null;
    }
};
exports.time = function () { return ({
    nowISO8601: function (t) {
        return moment().toISOString();
    },
    nowEpochSeconds: function () {
        return moment().unix();
    },
    nowEpochMilliSeconds: function () {
        return moment().valueOf();
    },
    nowFormatted: function (format, timezone) {
        if (timezone === void 0) { timezone = null; }
        try {
            if (timezone) {
                return moment()
                    .tz(timezone)
                    .formatWithJDF(format);
            }
            return moment().formatWithJDF(format);
        }
        catch (e) {
            return null;
        }
    },
    parseFormattedToEpochMilliSeconds: function (dateTime, format, timezone) {
        var timestamp = parseTimestamp(dateTime, format, timezone);
        return timestamp ? timestamp.valueOf() : null;
    },
    parseISO8601ToEpochMilliSeconds: function (dateTime) {
        var timestamp = parseTimestamp(dateTime, 'YYYY-MM-DDTHH:mm:ss.SZ');
        return timestamp ? timestamp.valueOf() : null;
    },
    epochMilliSecondsToSeconds: function (milliseconds) {
        try {
            return Math.floor(milliseconds / 1000);
        }
        catch (e) {
            return null;
        }
    },
    epochMilliSecondsToISO8601: function (dateTime) {
        try {
            return moment(dateTime).toISOString();
        }
        catch (e) {
            return null;
        }
    },
    epochMilliSecondsToFormatted: function (timestamp, format, timezone) {
        if (timezone === void 0) { timezone = 'UTC'; }
        try {
            return moment(timestamp)
                .tz(timezone)
                .formatWithJDF(format);
        }
        catch (e) {
            return null;
        }
    },
}); };
